### PR TITLE
[Suggestion] convertpcinfo script command

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: N/A

* **Server Mode**: -

* **Description of Pull Request**: 

## PRO
- Convert more easily a player data to another.
- Fail silently if the character is not found/online.
## CONS
- We can retrieve the same information using several commands if the player is online or we can store the information in some variable before usage.
## Usage
```
  convertpcinfo(<char_id>,<type>)
  convertpcinfo(<account_id>,<type>)
  convertpcinfo(<player_name>,<type>)
```
## Others
```
	mes "name from char ID : " +       convertpcinfo( getcharid(0), CPC_NAME );		// strcharinfo(0,<char_id>)
	mes "char ID from char ID : " +    convertpcinfo( getcharid(0), CPC_CHAR );		// -
	mes "account ID from char ID : " + convertpcinfo( getcharid(0), CPC_ACCOUNT );	// getcharid(3,strcharinfo(0,<char_id>))
	mes " ";
	mes "name from account ID : " +       convertpcinfo( getcharid(3), CPC_NAME );	// rid2name(<account_id>)
	mes "char ID from account ID : " +    convertpcinfo( getcharid(3), CPC_CHAR );	// getcharid(0,rid2name(<account_id>))
	mes "account ID from account ID : " + convertpcinfo( getcharid(3), CPC_ACCOUNT );	// -
	mes " ";
	mes "name from name : " +       convertpcinfo( strcharinfo(0), CPC_NAME );		// -
	mes "char ID from name : " +    convertpcinfo( strcharinfo(0), CPC_CHAR );		// getcharid(0,<name>)
	mes "account ID from name : " + convertpcinfo( strcharinfo(0), CPC_ACCOUNT );		// getcharid(3,<name>)
```

### CODE REVIEW

File: src/map/script.hpp [hunk 1]
```diff
@@ -1898,6 +1898,12 @@ enum e_hat_effects {
 	HAT_EF_MAX
 };
 
+enum bl_pc_type {
```
Comment (aleos89) [inline src/map/script.hpp]: Maybe like:
```suggestion
enum e_charinfo_type {
```

**Comment by sader1992** [comment]
i already used this command too many times , and i think it's great command.

File: src/map/script.cpp [hunk 2]
```diff
@@ -24026,6 +24026,50 @@ BUILDIN_FUNC(achievement_condition){
...
+*/
+BUILDIN_FUNC(charinfo) {
+	TBL_PC *sd;
+
+	if (script_isstring(st, 2))
+		sd = map_nick2sd(script_getstr(st, 2),false);
+	else {
+		int id = script_getnum(st, 2);
+		sd = map_id2sd(id);
+		if (sd == NULL)
```
Comment (Normynator) [inline src/map/script.cpp]: Redundant nullptr check

File: src/map/script.cpp [hunk 3]
```diff
@@ -24026,6 +24026,50 @@ BUILDIN_FUNC(achievement_condition){
...
+			sd = map_charid2sd(id);
+	}
+
+	switch( script_getnum(st, 3) ) {
+	case PC_NAME:
+		if (sd)
```
Comment (Normynator) [inline src/map/script.cpp]: The if-statements could be used outside.
Pseudocode:
`
if(sd) {
    Switch() ...
} else {
   Switch() ...
}`
But the compiler will optimize it that way anyway. So up to you.

**Comment by Atemo** [comment]
I have the feeling that the command name is not specific enough. Maybe `convert_pc_data`..? Any idea?

**Comment by sader1992** [comment]
what i see the name of the command is good , but maybe `pc_info` or `pcinfo`
and maybe later if we needed , we can expand the command to return more info for the pc.

but i still see `charinfo` is a good name , people should not read the command without it information.

File: src/map/script.cpp [hunk 4]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+		sd = map_nick2sd(script_getstr(st, 2),false);
+	else {
+		int id = script_getnum(st, 2);
+		sd = map_id2sd(id);
+		if (!sd)
+			sd = map_charid2sd(id);
+	}
+
+	int type = script_getnum(st, 3);
+	if (type < PC_NAME || type > PC_ACCOUNT) {
```
Comment (cydh) [inline src/map/script.cpp]: doesn't have multiline remove suggestion

```suggestion
```

File: src/map/script.cpp [hunk 5]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+		ShowError("buildin_charinfo: unknown type %d.\n", type);
```
Comment (cydh) [inline src/map/script.cpp]:
```suggestion
```

File: src/map/script.cpp [hunk 6]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+		script_pushnil(st);
```
Comment (cydh) [inline src/map/script.cpp]:
```suggestion
```

File: src/map/script.cpp [hunk 7]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+		st->state = END;
```
Comment (cydh) [inline src/map/script.cpp]:
```suggestion
```

File: src/map/script.cpp [hunk 8]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+		return SCRIPT_CMD_FAILURE;
```
Comment (cydh) [inline src/map/script.cpp]:
```suggestion
```

File: src/map/script.cpp [hunk 9]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+	}
```
Comment (cydh) [inline src/map/script.cpp]:
```suggestion
```

File: src/map/script.cpp [hunk 10]
```diff
@@ -24107,6 +24107,53 @@ BUILDIN_FUNC(achievement_condition){
...
+	case PC_NAME:
+		script_pushstrcopy(st, sd->status.name);
+		break;
+	case PC_CHAR:
+		script_pushint(st, sd->status.char_id);
+		break;
+	case PC_ACCOUNT:
+		script_pushint(st, sd->status.account_id);
+		break;
+	}
```
Comment (cydh) [inline src/map/script.cpp]:
```suggestion
	default:
		ShowError("buildin_charinfo: unknown type %d.\n", type);
		script_pushnil(st);
		st->state = END;
		return SCRIPT_CMD_FAILURE;
	}
```

**Pull request merged**